torch/ao/quantization/fx/_model_report/detector.py

```diff
@@ -0,0 +1,408 @@
+
+        # assert for MyPy check
+        assert isinstance(per_channel_info["per_channel_status"], dict)
+
+        optimizations_possible = False
+        for fqn in per_channel_info["per_channel_status"]:
+            fqn_dict = per_channel_info["per_channel_status"][fqn]
+            if fqn_dict["per_channel_supported"] and not fqn_dict["per_channel_used"]:
+                optimizations_possible = True
+                further_optims_str += "Module {module_fqn} can be configured to use per_channel quantization.\n".format(
+                    module_fqn=fqn
+                )
+
+        if optimizations_possible:
+            further_optims_str += (
+                "To use per_channel quantization, make sure the qconfig has a per_channel weight observer."
+            )
+        else:
+            further_optims_str += "No further per_channel optimizations possible."
+
+        # return the string and the dictionary form of same information
+        return (further_optims_str, per_channel_info)
+
+
+class DynamicStaticDetector(DetectorBase):
+    r"""
+    Determines whether dynamic or static quantization is more appropriate for a given module.
+
+    Takes advantage of the ModelReportObserver that records range information.
+    Stationary distribution of data are strictly above tolerance level for the comparison statistic:
+
+        S = average_batch_activation_range/epoch_activation_range
+
+    Nonstationary distributions are below or at the tolerance level for this metric.
+
+    If the distribution of data right after the module is non-stationary, recommend dynamic quantization
+        Otherwise recommend static quantization
+
+    Args:
+        tolerance (float, optional): The threshold where S metric is stationary above and non-stationary otherwise. Default: 0.5
+    """
+    # names for the pre and post observers that are inserted
+    DEFAULT_PRE_OBSERVER_NAME = "model_report_pre_observer"
+    DEFAULT_POST_OBSERVER_NAME = "model_report_post_observer"
+
+    # naming conventions for stationary vs non-stationary data
+    DEFAULT_STATIONARY = "stationary"
+    DEFAULT_NON_STATIONARY = "non-stationary"
+
+    # modules that are supported both dynamic and static for this report function
+    DEFAULT_DYNAMIC_STATIC_CHECK_SUPPORTED = set([nn.Linear])
+
+    def __init__(self, tolerance=0.5):
+        super().__init__()
+
+        # set tolerance level and initialize a set to keep track of useful fqn locations
+        self.tolerance = tolerance
+        self.useful_observer_fqns: Set[str] = set([])
+
+    def determine_observer_insert_points(self, model) -> Tuple[Set[str], Any]:
+
+        raise NotImplementedError("Will be implemented in a future commit")
+
+    def get_detector_name(self) -> str:
+        r""" returns the string name of this detector"""
+        return "dynamic_vs_static_detector"
+
+    def __generate_dict_info(self, model: GraphModule) -> Dict[str, Any]:
```

Comment:
should this be one _ or two __ ?

Reply:
I think it has to be 2 for it to be private, just tested it and if it's only 1, it still pops up as a callable method when we create an instance.

Reply:
looked up docs and I stand corrected, guess it should be one, i will fix